tests/unit/test_compress.py

```diff
@@ -0,0 +1,382 @@
+        and encoding in headers['Content-Encoding']
+    )
+
+
+def assert_compression(original_body, request_dict, encoding):
+    compressed_body = request_dict['body']
+    headers = request_dict['headers']
+    _assert_compression_body(original_body, compressed_body, encoding)
+    _assert_compression_header(headers, encoding)
+
+
+@pytest.mark.parametrize(
+    'config, request_dict, operation_model, encoding',
+    [
+        (
+            COMPRESSION_CONFIG_128_BYTES,
+            default_request_dict(),
+            OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+        (
+            COMPRESSION_CONFIG_128_BYTES,
+            default_request_dict(),
+            OP_MULTIPLE_COMPRESSIONS,
+            'gzip',
+        ),
+        (
+            DEFAULT_COMPRESSION_CONFIG,
+            default_request_dict(),
+            STREAMING_OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+        (
+            COMPRESSION_CONFIG_128_BYTES,
+            request_dict_bytearray(),
+            OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+        (
+            COMPRESSION_CONFIG_128_BYTES,
+            request_dict_with_content_encoding_header(),
+            OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+        (
+            COMPRESSION_CONFIG_128_BYTES,
+            request_dict_string(),
+            OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+        (
+            COMPRESSION_CONFIG_128_BYTES,
+            request_dict_bytes_io(),
+            OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+        (
+            COMPRESSION_CONFIG_128_BYTES,
+            request_dict_string_io(),
+            OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+        (
+            COMPRESSION_CONFIG_1_BYTE,
+            request_dict_dict(),
+            OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+    ],
+)
+def test_compression(config, request_dict, operation_model, encoding):
+    original_body = request_dict['body']
+    maybe_compress_request(config, request_dict, operation_model)
+    assert_compression(original_body, request_dict, encoding)
+
+
+@pytest.mark.parametrize(
+    'config, request_dict, operation_model, encoding',
+    [
+        (
+            DEFAULT_COMPRESSION_CONFIG,
+            request_dict_non_seekable_bytes_stream(),
+            STREAMING_OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+        (
+            DEFAULT_COMPRESSION_CONFIG,
+            request_dict_non_seekable_text_stream(),
+            STREAMING_OP_WITH_COMPRESSION,
+            'gzip',
+        ),
+    ],
+)
+def test_compression_non_seekable_streams(
+    config, request_dict, operation_model, encoding
+):
+    # since the body can't be reset, we must make a copy
+    # of the original body to test against
+    original_body = deepcopy(request_dict['body'])
+    maybe_compress_request(config, request_dict, operation_model)
+    assert_compression(original_body, request_dict, encoding)
```

Comment:
This can definitely be rolled up into the normal `test_compression` test cases, but it didn't seem right to make copies of the body when it isn't necessary.